### PR TITLE
[codegen][Python] Add model cache to speed up code generation

### PR DESCRIPTION
**Pull request by sebastien-rosset**
This is a performance enhancement. The code generation behavior does not change:

1. Create a cache for `StringUtils.escape()`
1. Create a cache from named model to CodegenModel. Used by `fromModel()`
1. Create a cache from named property to CodegenProperty. Used by `fromProperty()`
1. Create a cache from language-specific model name to Shema.
1. Pre-compile regex patterns instead of calling String.replaceAll().
1. Other minor optimizations

I've noticed for a large OpenAPI doc, the code generation was taking more than 3 minutes and 47% of all the CPU processing time was spent invoking the `findFirst()` method in PythonClientExperimentalCodegen.java (as shown below). This PR creates a map from toModelName() to schema such that toModelName() does not have to be called many time repeatedly and the Schema map does not have to be traversed repeatedly.

```java
 Optional<Schema> referencedSchema = ModelUtils.getSchemas(openAPI).entrySet().stream()
                .filter(entry -> Objects.equals(varDataType, toModelName(entry.getKey())))
                .map(Map.Entry::getValue)
                .findFirst();
```

https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java#L453

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) beforehand.
- [x] Run the shell script `./bin/generate-samples.sh`to update all Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. These must match the expectations made by your contribution. You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### CODE REVIEW

**Comment by wing328** [comment]
cc @OpenAPITools/generator-core-team

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java [hunk 1]
```diff
@@ -449,6 +452,23 @@ private void registerMustacheLambdas() {
         return objs;
     }
 
+    /**
+     * Return a map from model name to Schema for efficient lookup.
+     * 
+     * @return map from model name to Schema.
+     */
+    protected Map<String, Schema> getModelNameToSchemaCache() {
```
Comment (spacether) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:460]: Are there any locations where DefaultCodegen.java should use this?

**Comment by spacether** [comment]
@sebastien-rosset `taking more than 3 minutes and 47% of all the CPU processing time was spent invoking the findFirst()`
What is the performance improvement with this update?

**Comment by sebastien-rosset** [comment]
Other hot spots identified with possible optimization opportunities:

* `fromModel()`
* `fromProperty()`
* `createDiscriminator()`
* `toVarName()`
* `toParamName()`

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java (hunk 1 above)
Comment (spacether) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:460]: Thank you for doing that!

**Comment by spacether** [comment]
If fixing all of the CI errors is too much for this one PR, maybe you should separate these performance optimizations into multiple PRs.

**Comment by sebastien-rosset** [comment]
No, it should be a simple change.

**Comment by sebastien-rosset** [comment]
> @sebastien-rosset `taking more than 3 minutes and 47% of all the CPU processing time was spent invoking the findFirst()`
> What is the performance improvement with this update?

With the various enhancements so far, there is a 9.6x performance improvement for one large OAS document I have tested. There are multiple phases and options in the code generation (e.g. parsing, generation...), so the actual improvement depends on the command-line arguments.

**Comment by sebastien-rosset** [comment]
After these optimizations the next hot spot is createDiscriminator, which takes 28% CPU time when running codegen under a profiler.

**Comment by sebastien-rosset** [comment]
Unrelated build problem:
note: Unresolved dependencies path:
[error] sbt.librarymanagement.ResolveException: Error downloading com.fasterxml.jackson.core:jackson-annotations:2.9.10

**Comment by spacether** [comment]
> After these optimizations the next hot spot is createDiscriminator, which takes 28% CPU time when running codegen under a profiler.

Yup, that one does a buunch of recursive traversal of schemas to check for discriminators.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/StringUtils.java [hunk 2]
```diff
@@ -57,18 +75,18 @@
      * @return The underscored version of the word
      */
     public static String underscore(final String word) {
-        return underscoreWords.get(word, wordToUnderscore -> {
+        return underscoreWordsCache.get(word, wordToUnderscore -> {
             String result;
             String firstPattern = "([A-Z]+)([A-Z][a-z])";
             String secondPattern = "([a-z\\d])([A-Z])";
```
Comment (spacether) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/StringUtils.java:81]: How about deleting firstPattern + secondPattern now that you are no longer using them?

Comment (sebastien-rosset) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/StringUtils.java]: Thanks for catching this. Done

**Comment by sebastien-rosset** [comment]
I found a way to further improve performance by another 20%, but since it's a bit more involved, I am proposing to submit as a separate PR.

**Comment by spacether** [comment]
> I found a way to further improve performance by another 20%, but since it's a bit more involved, I am proposing to submit as a separate PR.

Thank you for keeping it compartmentalized. That way we get most of the fixes now; unblock my PR and add that further improvement later.

**Comment by sebastien-rosset** [comment]
Build failed with unrelated issue:
```
[ERROR] Failed to execute goal on project scala-akka-petstore-client: Could not resolve dependencies for project org.openapitools:scala-akka-petstore-client:jar:1.0.0: Could not transfer artifact com.typesafe.akka:akka-stream_2.12:jar:2.5.21 from/to central (https://repo.maven.apache.org/maven2): Connect to repo.maven.apache.org:443 [repo.maven.apache.org/199.232.64.215] failed: Connection timed out (Connection timed out) -> [Help 1]

**Comment by sebastien-rosset** [comment]
Thanks!

**Comment by wing328** [comment]
Looks like this has impacts on other generators, e.g.
```diff
+++ b/samples/client/petstore/haskell-http-client/lib/OpenAPIPetstore/Model.hs
@@ -900,7 +900,7 @@ mkClassModel =
 -- ** Client
 -- | Client
 data Client = Client
-  { clientClient :: !(Maybe Text) -- ^ "client"
+  { clientClientClientClientClientClientClientClient :: !(Maybe Text) -- ^ "client"
   } deriving (P.Show, P.Eq, P.Typeable)
 
 -- | FromJSON Client
@@ -913,7 +913,7 @@ instance A.FromJSON Client where
 instance A.ToJSON Client where
   toJSON Client {..} =
    _omitNulls
-      [ "client" .= clientClient
+      [ "client" .= clientClientClientClientClientClientClientClient
       ]
 
 
@@ -922,7 +922,7 @@ mkClient
   :: Client
 mkClient =
   Client
-  { clientClient = Nothing
+  { clientClientClientClientClientClientClientClient = Nothing
   }
 
 -- ** Dog
@@ -1099,8 +1099,8 @@ mkFile =
 -- ** FileSchemaTestClass
 -- | FileSchemaTestClass
 data FileSchemaTestClass = FileSchemaTestClass
-  { fileSchemaTestClassFile :: !(Maybe File) -- ^ "file"
-  , fileSchemaTestClassFiles :: !(Maybe [File]) -- ^ "files"
+  { fileSchemaTestClassFileSchemaTestClassFile :: !(Maybe File) -- ^ "file"
+  , fileSchemaTestClassFileSchemaTestClassFiles :: !(Maybe [File]) -- ^ "files"
   } deriving (P.Show, P.Eq, P.Typeable)
 

```

I'll revert it for the time being.

**Comment by spacether** [comment]
Thank you

**Comment by wing328** [comment]
Reverted via #7245 

@sebastien-rosset please open a new PR when you've time and we'll proceed from there.

**Comment by sebastien-rosset** [comment]
> Looks like this has impacts on other generators, e.g.
> 
> ```diff
> +++ b/samples/client/petstore/haskell-http-client/lib/OpenAPIPetstore/Model.hs
> @@ -900,7 +900,7 @@ mkClassModel =
>  -- ** Client
>  -- | Client
>  data Client = Client
> -  { clientClient :: !(Maybe Text) -- ^ "client"
> +  { clientClientClientClientClientClientClientClient :: !(Maybe Text) -- ^ "client"
>    } deriving (P.Show, P.Eq, P.Typeable)
>  
>  -- | FromJSON Client
> @@ -913,7 +913,7 @@ instance A.FromJSON Client where
>  instance A.ToJSON Client where
>    toJSON Client {..} =
>     _omitNulls
> -      [ "client" .= clientClient
> +      [ "client" .= clientClientClientClientClientClientClientClient
>        ]
>  
>  
> @@ -922,7 +922,7 @@ mkClient
>    :: Client
>  mkClient =
>    Client
> -  { clientClient = Nothing
> +  { clientClientClientClientClientClientClientClient = Nothing
>    }
>  
>  -- ** Dog
> @@ -1099,8 +1099,8 @@ mkFile =
>  -- ** FileSchemaTestClass
>  -- | FileSchemaTestClass
>  data FileSchemaTestClass = FileSchemaTestClass
> -  { fileSchemaTestClassFile :: !(Maybe File) -- ^ "file"
> -  , fileSchemaTestClassFiles :: !(Maybe [File]) -- ^ "files"
> +  { fileSchemaTestClassFileSchemaTestClassFile :: !(Maybe File) -- ^ "file"
> +  , fileSchemaTestClassFileSchemaTestClassFiles :: !(Maybe [File]) -- ^ "files"
>    } deriving (P.Show, P.Eq, P.Typeable)
> ```
> 
> List of impacted files:
> 
> ```
> Changes not staged for commit:
>   (use "git add <file>..." to update what will be committed)
>   (use "git checkout -- <file>..." to discard changes in working directory)
> 
> 	modified:   samples/client/petstore/haskell-http-client/lib/OpenAPIPetstore/Model.hs
> 	modified:   samples/client/petstore/haskell-http-client/lib/OpenAPIPetstore/ModelLens.hs
> 	modified:   samples/client/petstore/haskell-http-client/tests/Instances.hs
> 	modified:   samples/openapi3/client/petstore/python-experimental/docs/FakeApi.md
> 	modified:   samples/openapi3/client/petstore/python-experimental/petstore_api/api/fake_api.py
> ```
> 
> I'll revert it for the time being.

Just curious, I thought the build process was failing when the workspace has locally modified files?

**Comment by wing328** [comment]
@sebastien-rosset yes the build process (circieci) failed: https://circleci.com/gh/OpenAPITools/openapi-generator/18334?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Pull request merged**